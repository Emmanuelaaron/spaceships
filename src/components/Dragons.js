import { useSelector, useDispatch } from 'react-redux';
import { useEffect } from 'react';
import { bindActionCreators } from 'redux';
import Card from 'react-bootstrap/Card';
import Button from 'react-bootstrap/Button';
import Badge from 'react-bootstrap/Badge';
import { loadDragons, reserveDragon, cancelDragonReservation } from '../redux/dragons/dragons';

const Dragons = () => {
  const dispatch = useDispatch();
  const loadDragonsAction = bindActionCreators(loadDragons, dispatch);
  const reserveDragonAction = bindActionCreators(reserveDragon, dispatch);
  const cancelDragonReservationAction = bindActionCreators(cancelDragonReservation, dispatch);
  const dragons = useSelector((state) => state.dragons);

  useEffect(() => {
    if (dragons.length === 0) loadDragonsAction();
    return () => null;
  }, []);

  return (
    <div className="d-flex flex-column flex-md-row mt-5 container align-items-md-start align-items-center align-items-md-stretch">
      {dragons.map((dragon) => (
        <Card
          bg="light"
          key={dragon.id}
          text="dark"
          style={{ width: '50vw', minWidth: '20rem' }}
          className="m-2"
        >
          <Card.Img
            variant="left"
            src={dragon.flickr_images[0]}
            style={{
              width: '100%',
              height: '100%',
              objectFit: 'cover',
            }}
          />
          <Card.Header>
            {dragon.type.charAt(0).toUpperCase() + dragon.type.slice(1)}
            {' '}
            {dragon.reserved && (
              <Badge bg="success">Reserved</Badge>
            )}
          </Card.Header>
          <Card.Body className="d-flex flex-column align-items-center justify-content-center">
            <Card.Title>{dragon.name}</Card.Title>
            <Button
              variant={dragon.reserved ? 'outline-secondary' : 'primary'}
              onClick={
                () => {
                  if (dragon.reserved) {
                    cancelDragonReservationAction(dragon.id);
                  } else reserveDragonAction(dragon.id);
                }
              }
            >
              {dragon.reserved ? 'Cancel Reservation' : 'Reserve Dragon'}
            </Button>
          </Card.Body>
        </Card>
      ))}
    </div>
  );
};

export default Dragons;
