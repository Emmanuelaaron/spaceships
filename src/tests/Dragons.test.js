import { Provider } from 'react-redux';
import { render, screen } from '@testing-library/react';
import store from '../redux/configureStore';
import Dragons from '../components/Dragons';

describe('Dragons component', () => {
  describe('Use Jest snapshots to test the Dragons component', () => {
    it('renders Dragons component as expected', async () => {
      const dragonsComponent = render(
        <Provider store={store}>
          <Dragons />
        </Provider>,
      );

      expect(await screen.findAllByText('Reserve Dragon')).toBeTruthy();
      expect(dragonsComponent.asFragment()).toMatchSnapshot();
    });
  });
});
