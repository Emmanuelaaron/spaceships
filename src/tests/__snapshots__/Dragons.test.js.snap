// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dragons component Use Jest snapshots to test the Dragons component renders Dragons component as expected 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-column flex-md-row mt-5 container align-items-md-start align-items-center align-items-md-stretch"
  >
    <div
      class="m-2 card bg-light text-dark"
      style="width: 50vw; min-width: 20rem;"
    >
      <img
        class="card-img-left"
        src="https://i.imgur.com/9fWdwNv.jpg"
        style="width: 100%; height: 100%; object-fit: cover;"
      />
      <div
        class="card-header"
      >
        Capsule 
      </div>
      <div
        class="d-flex flex-column align-items-center justify-content-center card-body"
      >
        <div
          class="card-title h5"
        >
          Dragon 1
        </div>
        <button
          class="btn btn-primary"
          type="button"
        >
          Reserve Dragon
        </button>
      </div>
    </div>
    <div
      class="m-2 card bg-light text-dark"
      style="width: 50vw; min-width: 20rem;"
    >
      <img
        class="card-img-left"
        src="https://farm8.staticflickr.com/7647/16581815487_6d56cb32e1_b.jpg"
        style="width: 100%; height: 100%; object-fit: cover;"
      />
      <div
        class="card-header"
      >
        Capsule 
      </div>
      <div
        class="d-flex flex-column align-items-center justify-content-center card-body"
      >
        <div
          class="card-title h5"
        >
          Dragon 2
        </div>
        <button
          class="btn btn-primary"
          type="button"
        >
          Reserve Dragon
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
